cpp/doc/libcuspatial_refactoring_guide.md

```diff
@@ -0,0 +1,272 @@
+# cuSpatial C++ API Refactoring Guide
+
+The original cuSpatial C++ API (libcuspatial) was designed to depend on RAPIDS libcudf and use
+its core data types, especially `cudf::column`. For users who do not also use libcudf or other 
+RAPIDS APIS, depending on libcudf could be a big barrier to adoption of libcuspatial. libcudf is 
+a very large library and building it takes a lot of time.
+
+Therefore, we are developing a standalone libcuspatial C++ API that does not depend on libcudf. This 
+is a header-only template API with an iterator-based interface. This has a number of advantages
+
+  1. With a header-only API, users can include and build exactly what they use.
+  2. With a templated API, the API can be flexible to support a variety of basic data types, such 
+     as float and double for positional data, and different integer sizes for indices.
+  3. By templating on iterator types, cuSpatial algorithms can be fused with transformations of the
+     input data, by using "fancy" iterators. Examples include transform iterators and counting
+     iterators.
+  4. Memory resources only need to be part of APIs that allocate temporary intermediate storage.
+     Output storage is allocated outside the API and an output iterator is passed as an argument. 
+
+The main disadvantages of this type of API are
+
+  1. Header-only APIs can increase compilation time for code that depends on them.
+  2. Some users (especially our Python API) may prefer a cuDF-based API.
+
+The good news is that by maintaining the existing libcudf-based C++ API as a layer above the header-
+only libcuspatial API, we can avoid problem 1 and problem 2 for users of the legacy API.
+
+## Example API
+
+Following is an example iterator-based API for `cuspatial::haversine_distance`.
+
+```C++
+template <class LonLatItA,
+          class LonLatItB,
+          class OutputIt,
+          class Location = typename std::iterator_traits<LonLatItA>::value_type,
+          class T        = typename Location::value_type>
+OutputIt haversine_distance(LonLatItA a_lonlat_first,
+                            LonLatItA a_lonlat_last,
+                            LonLatItB b_lonlat_first,
+                            OutputIt distance_first,
+                            T const radius = EARTH_RADIUS_KM,
+                            rmm::cuda_stream_view stream = rmm::cuda_stream_default);
+```
+
+There are a few key points to notice.
+
+  1. The API is very similar to STL algorithms such as `std::transform`.
+  2. All array inputs and outputs are iterator type templates. 
+  3. Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d`
+     type (include/cuspatial/types.hpp)
```

Comment:
The code doesn't currently tell me this. Maybe add a `static_assert` to demonstrate it.

Reply:
Concepts would be so nice for this :(

Reply:
I don't understand, I already have that static assert. 

```
static_assert(
    std::conjunction_v<std::is_same<lonlat_2d<T>, Location>, std::is_same<lonlat_2d<T>, LocationB>>,
    "Inputs must be cuspatial::lonlat_2d");
```

Reply:
Mentioned the static asserts in the refactoring guide.

Reply:
Reading the code on lines33-43 and then reading on it describes the important things about this code...

"There are a few key points to notice...

Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d` type"

My only point is that it's impossible to "notice" this point just from the code as is on 33-43.

Reply:
Got it. I added this clarification to that point: 

```
 Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d`
     type (include/cuspatial/types.hpp). This is enforced using a `static_assert` in the function
     body (discussed later).
```